File: ReactJS-Spring-Boot-CRUD-Full-Stack-App/react-frontend/src/services/EmployeeService.js
import axiosClient from '../api/axiosClient';

class EmployeeService {
    getEmployees({ page = 0, size = 10, sortBy = 'id', sortDirection = 'ASC' } = {}) {
        return axiosClient.get('/employees', { params: { page, size, sortBy, sortDirection } });
    }

    searchEmployees(keyword, { page = 0, size = 10 } = {}) {
        return axiosClient.get('/employees/search', { params: { keyword, page, size } });
    }

    filterEmployees(filters = {}, { page = 0, size = 10 } = {}) {
        return axiosClient.get('/employees/filter', { params: { ...filters, page, size } });
    }

    getEmployeeById(id) {
        return axiosClient.get(`/employees/${id}`);
    }

    getMyProfile() {
        return axiosClient.get('/employees/me');
    }

    createEmployee(employee) {
        return axiosClient.post('/employees', employee);
    }

    updateEmployee(id, employee) {
        return axiosClient.put(`/employees/${id}`, employee);
    }

    deleteEmployee(id) {
        return axiosClient.delete(`/employees/${id}`);
    }
}

export default new EmployeeService();

File: ReactJS-Spring-Boot-CRUD-Full-Stack-App/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java
package net.javaguides.springboot.service;

import net.javaguides.springboot.dto.EmployeeRequestDto;
import net.javaguides.springboot.dto.EmployeeResponseDto;
import net.javaguides.springboot.dto.PageResponseDto;
import net.javaguides.springboot.entity.enums.Department;
import net.javaguides.springboot.entity.enums.EmployeeStatus;

/**
 * Business operations for Employee. Controllers depend on this interface,
 * not on the repository directly, so HTTP concerns stay separate from
 * business logic and transaction boundaries.
 */
public interface EmployeeService {

    PageResponseDto<EmployeeResponseDto> getAllEmployees(int page, int size, String sortBy, String sortDirection);

    PageResponseDto<EmployeeResponseDto> searchEmployees(String keyword, int page, int size);

    PageResponseDto<EmployeeResponseDto> filterEmployees(String name, String email, Department department,
                                                           String designation, EmployeeStatus status,
                                                           int page, int size);

    EmployeeResponseDto getEmployeeById(Long id);

    EmployeeResponseDto getEmployeeByEmail(String email);

    EmployeeResponseDto createEmployee(EmployeeRequestDto employeeRequestDto);

    EmployeeResponseDto updateEmployee(Long id, EmployeeRequestDto employeeRequestDto);

    void deleteEmployee(Long id);
}
